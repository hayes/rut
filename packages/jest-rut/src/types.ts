/* eslint-disable */

import React from 'react';

declare global {
  namespace jest {
    interface Matchers<R> {
      toBeElementType(type: React.ReactType): R;
      toContainNode(node: NonNullable<React.ReactNode>): R;
      toRenderChildren(): R;
    }
  }
}
