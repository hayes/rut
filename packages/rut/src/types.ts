/* eslint-disable @typescript-eslint/no-namespace */

import React from 'react';

export type Args<T> = T extends (...args: unknown[]) => unknown ? Parameters<T> : unknown[];

export interface UnknownProps {
  [name: string]: unknown;
}

export interface TestNode {
  children: (TestNode | string)[];
  instance: unknown;
  parent: TestNode | null;
  props: UnknownProps;
  type: React.ElementType;
}

export interface FiberNode {
  child: FiberNode | null;
  elementType: React.ElementType;
  index: number;
  key: string | null;
  mode: number;
  ref: React.RefObject<unknown> | null;
  return: FiberNode | null;
  sibling: FiberNode | null;
  tag: number;
  type: React.ElementType;
}

export interface MatchResult {
  context?: string;
  message: string;
  notMessage: string;
  passed: boolean;
}

export type NodeType =
  | 'class-component'
  // | 'context-consumer'
  // | 'context-provider'
  | 'forward-ref'
  | 'function-component'
  | 'host-component'
  | 'indeterminate-component'
  | 'fragment'
  | 'lazy'
  | 'memo'
  | 'mode'
  | 'portal'
  | 'profiler'
  | 'root'
  | 'suspense'
  | 'text';

declare module 'react-test-renderer' {
  interface ReactTestInstance {
    _fiber: FiberNode;
  }
}

declare global {
  namespace jest {
    interface Matchers<R> {
      toBeElementType(type: React.ElementType): R;
      toBeNodeType(type: NodeType): R;
      toContainNode(node: NonNullable<React.ReactNode>): R;
      toRenderChildren(): R;
    }
  }
}
