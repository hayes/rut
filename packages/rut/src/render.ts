import React from 'react';
import { act } from 'react-test-renderer';
import Renderer from './Renderer';
import { RendererOptions } from './types';

export default function render<Props>(
  element: React.ReactElement<Props>,
  options?: RendererOptions,
): Renderer<Props> {
  let renderer: Renderer<Props>;

  act(() => {
    renderer = new Renderer(element, options);
  });

  return renderer!;
}
