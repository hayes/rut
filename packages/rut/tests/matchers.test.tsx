import React, { Suspense, Profiler } from 'react';
// import ReactDOM from 'react-dom';
import render from '../src/render';
import { NodeType } from '../lib/types';

describe('matchers', () => {
  function MismatchComp() {
    return null;
  }

  function FuncComp() {
    return null;
  }

  class ClassComp extends React.Component {
    render() {
      return null;
    }
  }

  const ForwardRefComp = React.forwardRef(FuncComp);
  const MemoComp = React.memo(FuncComp);

  describe('toBeElementType()', () => {
    it('errors if a non-Element is passed', () => {
      expect(() => {
        expect(123).toBeElementType('div');
      }).toThrowError('toBeElementType: Expected a Rut `Element`.');
    });

    describe('host components', () => {
      it('passes when types match', () => {
        expect(() => {
          expect(render(<div />).root()).toBeElementType('div');
        }).not.toThrowError();
      });

      it('errors when types dont match', () => {
        expect(() => {
          expect(render(<div />).root()).toBeElementType('span');
        }).toThrowError('expected `div` to be a `span`');
      });

      it('passes when types match (not negation)', () => {
        expect(() => {
          expect(render(<div />).root()).not.toBeElementType('span');
        }).not.toThrowError();
      });

      it('errors when types dont match (not negation)', () => {
        expect(() => {
          expect(render(<div />).root()).not.toBeElementType('div');
        }).toThrowError('expected `div` not to be a `div`');
      });
    });

    describe('function components', () => {
      it('passes when types match', () => {
        expect(() => {
          expect(render(<FuncComp />).root()).toBeElementType(FuncComp);
        }).not.toThrowError();
      });

      it('errors when types dont match', () => {
        expect(() => {
          expect(render(<FuncComp />).root()).toBeElementType(MismatchComp);
        }).toThrowError('expected `FuncComp` to be a `MismatchComp`');
      });

      it('passes when types match (not negation)', () => {
        expect(() => {
          expect(render(<FuncComp />).root()).not.toBeElementType(MismatchComp);
        }).not.toThrowError();
      });

      it('errors when types dont match (not negation)', () => {
        expect(() => {
          expect(render(<FuncComp />).root()).not.toBeElementType(FuncComp);
        }).toThrowError('expected `FuncComp` not to be a `FuncComp`');
      });
    });

    describe('class components', () => {
      it('passes when types match', () => {
        expect(() => {
          expect(render(<ClassComp />).root()).toBeElementType(ClassComp);
        }).not.toThrowError();
      });

      it('errors when types dont match', () => {
        expect(() => {
          expect(render(<ClassComp />).root()).toBeElementType(MismatchComp);
        }).toThrowError('expected `ClassComp` to be a `MismatchComp`');
      });

      it('passes when types match (not negation)', () => {
        expect(() => {
          expect(render(<ClassComp />).root()).not.toBeElementType(MismatchComp);
        }).not.toThrowError();
      });

      it('errors when types dont match (not negation)', () => {
        expect(() => {
          expect(render(<ClassComp />).root()).not.toBeElementType(ClassComp);
        }).toThrowError('expected `ClassComp` not to be a `ClassComp`');
      });
    });
  });

  describe('toBeNodeType()', () => {
    it('errors if a non-Element is passed', () => {
      expect(() => {
        expect(123).toBeNodeType('function-component');
      }).toThrowError('toBeNodeType: Expected a Rut `Element`.');
    });

    it('errors if an invalid node type', () => {
      expect(() => {
        expect(render(<div />).root()).toBeNodeType(
          // @ts-ignore Allow invalid
          'unknown-type',
        );
      }).toThrowErrorMatchingSnapshot();
    });

    const nodeTypeMap = {
      'class-component': <ClassComp />,
      // 'context-consumer': <Ctx.Consumer>{() => null}</Ctx.Consumer>,
      // 'context-provider': <Ctx.Provider value="test">{null}</Ctx.Provider>,
      'forward-ref': <ForwardRefComp />,
      fragment: (
        <>
          <div />
        </>
      ),
      'function-component': <FuncComp />,
      'host-component': <div />,
      lazy: React.lazy(() => Promise.resolve({ default: FuncComp })),
      memo: <MemoComp />,
      // mode: 8,
      //      portal: ReactDOM.createPortal(<div />),
      profiler: <Profiler id="test" onRender={jest.fn()} />,
      // root: 3,
      suspense: <Suspense fallback={<div />} />,
      text: 'test',
    };

    Object.entries(nodeTypeMap).forEach(([testTypeName, testNode]) => {
      const wrapper = render(<div>{testNode}</div>);
      const expectedNode = wrapper.root().children()[0];
      const typeName = testTypeName as NodeType;

      describe(`${typeName} type`, () => {
        it('passes when types match', () => {
          expect(() => {
            expect(expectedNode).toBeNodeType(typeName);
          }).not.toThrowError();
        });

        it('errors when types dont match', () => {
          expect(() => {
            expect(expectedNode).toBeNodeType(typeName === 'root' ? 'mode' : 'root');
          }).toThrowErrorMatchingSnapshot();
        });

        it('passes when types match (not negation)', () => {
          expect(() => {
            expect(expectedNode).not.toBeNodeType(typeName === 'root' ? 'mode' : 'root');
          }).not.toThrowError();
        });

        it('errors when types dont match (not negation)', () => {
          expect(() => {
            expect(expectedNode).not.toBeNodeType(typeName);
          }).toThrowErrorMatchingSnapshot();
        });
      });
    });
  });
});
