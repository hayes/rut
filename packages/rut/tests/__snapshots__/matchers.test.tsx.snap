// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchers toBeNodeType() class-component type errors when types dont match (not negation) 1`] = `"expected \`ClassComp\` not to be a \\"class-component\\""`;

exports[`matchers toBeNodeType() class-component type errors when types dont match 1`] = `"expected \`ClassComp\` to be a \\"root\\""`;

exports[`matchers toBeNodeType() context-consumer type errors when types dont match (not negation) 1`] = `"toBeNodeType: Expected a Rut \`Element\`."`;

exports[`matchers toBeNodeType() context-consumer type errors when types dont match 1`] = `"toBeNodeType: Expected a Rut \`Element\`."`;

exports[`matchers toBeNodeType() errors if an invalid node type 1`] = `"toBeNodeType: Invalid node type \\"unknown-type\\", expected one of: class-component, context-consumer, context-provider, forward-ref, fragment, function-component, host-component, indeterminate-component, lazy, memo, mode, portal, profiler, root, suspense, text"`;

exports[`matchers toBeNodeType() forward-ref type errors when types dont match (not negation) 1`] = `"Cannot convert object to primitive value"`;

exports[`matchers toBeNodeType() forward-ref type errors when types dont match 1`] = `"Cannot convert object to primitive value"`;

exports[`matchers toBeNodeType() fragment type errors when types dont match 1`] = `"expected \`div\` to be a \\"root\\""`;

exports[`matchers toBeNodeType() function-component type errors when types dont match (not negation) 1`] = `"expected \`FuncComp\` not to be a \\"function-component\\""`;

exports[`matchers toBeNodeType() function-component type errors when types dont match 1`] = `"expected \`FuncComp\` to be a \\"root\\""`;

exports[`matchers toBeNodeType() host-component type errors when types dont match (not negation) 1`] = `"expected \`div\` not to be a \\"host-component\\""`;

exports[`matchers toBeNodeType() host-component type errors when types dont match 1`] = `"expected \`div\` to be a \\"root\\""`;
